### PR TITLE
feat: add streak config

### PR DESCRIPTION
**Pull request by omBratteng**
Related apps PR https://github.com/dailydotdev/apps/pull/3386

AS-437

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -9,6 +9,18 @@ module.exports = {
   rules: {
     'prettier/prettier': 'error',
     'unused-imports/no-unused-imports': 'error',
+    'no-restricted-imports': [
```
Comment (rebelchris) [inline .eslintrc.js:12]: Nice thanks for adding this failsafe :D

**Review by rebelchris — APPROVED** [review]
Looks good from my side 🙏

File: src/routes/boot.ts [hunk 2]
```diff
@@ -383,6 +383,7 @@ const getUser = (con: DataSource, userId: string): Promise<User | null> =>
       'createdAt',
       'cover',
       'experienceLevel',
+      'weekStart',
```
Comment (capJavert) [inline src/routes/boot.ts:386]: do we need it in boot?

File: src/cron/hourlyNotifications.ts [hunk 3]
```diff
@@ -24,7 +26,7 @@ const cron: Cron = {
   handler: async (con, logger) => {
     const personalizedDigestQuery = con
       .createQueryBuilder()
-      .select('upd.*, u.timezone')
+      .select('upd.*, u.timezone, u.weekStart')
```
Comment (capJavert) [inline src/cron/hourlyNotifications.ts:29]: think it needs `u."weekStarts"` due to casing

File: src/common/users.ts [hunk 4]
```diff
@@ -406,7 +408,20 @@ export const clearUserStreak = async (
...
+  startOfWeek: DayOfWeek = DEFAULT_WEEK_START,
+) => {
+  if (startOfWeek === DayOfWeek.Sunday) {
+    return (
+      (day === Day.Saturday && difference > FREEZE_DAYS_IN_A_WEEK) ||
+      (day === Day.Sunday &&
+        difference > FREEZE_DAYS_IN_A_WEEK + MISSED_LIMIT) ||
+      (day > Day.Sunday && difference > MISSED_LIMIT)
+    );
+  }
```
Comment (capJavert) [inline src/common/users.ts:423]: :fml: 😆 a bit hard to parse, but it is what iti s

File: src/cron/dailyDigest.ts [hunk 5]
```diff
@@ -25,7 +27,7 @@ const cron: Cron = {
   handler: async (con, logger) => {
     const personalizedDigestQuery = con
       .createQueryBuilder()
-      .select('upd.*, u.timezone')
+      .select('upd.*, u.timezone, u.weekStart')
```
Comment (capJavert) [inline src/cron/dailyDigest.ts:30]: same as https://github.com/dailydotdev/daily-api/pull/2092#discussion_r1701642228

File: src/common/users.ts (hunk 4 above)
Comment (omBratteng) [reply src/common/users.ts:423]: I know 😢
Tried to think of a better way to solve it, but alas to no avail. Maybe worth to take a look another time.

File: src/routes/boot.ts (hunk 2 above)
Comment (omBratteng) [reply src/routes/boot.ts:386]: Technically not needed in boot, but we need it when we render the streak modal.
Same on notifications settings page

File: src/cron/hourlyNotifications.ts (hunk 3 above)
Comment (omBratteng) [reply src/cron/hourlyNotifications.ts]: Not sure, but I added just to be safe

File: .eslintrc.js (hunk 1 above)
Comment (omBratteng) [reply .eslintrc.js:12]: Yeah, thought it was a good addition just to be safe that we don't accidentally import `isWeekend` from `date-fns` and expect it to work differently.
Actually surprise that `isWeekend` from `date-fns` didn't have this feature...

Comment (capJavert) [reply .eslintrc.js:12]: Even worse they have it on `startOfWeek` util 🙈

File: src/routes/boot.ts (hunk 2 above)
Comment (capJavert) [reply src/routes/boot.ts:386]: If its not used anywhere from boot I'd remove it, because streak we pull from separate query I think?

**Review by capJavert — APPROVED** [review]

File: src/routes/boot.ts (hunk 2 above)
Comment (omBratteng) [reply src/routes/boot.ts:386]: I'll see if I can add it to `userStreak` query, and not in boot.

**Comment by omBratteng** [comment]
@capJavert refactored it out of boot and into streak query. Working on refactoring in apps PR.

File: src/common/users.ts (hunk 4 above)
Comment (omBratteng) [reply src/common/users.ts:423]: @capJavert what do you think about this? A little bit more readable hopefully

```ts
export const shouldResetStreak = (
  day: number,
  difference: number,
  startOfWeek: DayOfWeek = DEFAULT_WEEK_START,
) => {
  const firstDayOfWeek =
    startOfWeek === DayOfWeek.Monday ? Day.Monday : Day.Sunday;

  const lastDayOfWeek =
    startOfWeek === DayOfWeek.Monday ? Day.Sunday : Day.Saturday;

  return (
    (day === lastDayOfWeek && difference > FREEZE_DAYS_IN_A_WEEK) ||
    (day === firstDayOfWeek &&
      difference > FREEZE_DAYS_IN_A_WEEK + MISSED_LIMIT) ||
    (day > firstDayOfWeek && difference > MISSED_LIMIT)
  );
};
```

**Review by rebelchris — APPROVED** [review]
Approved.
Personally find this one better indeed: https://github.com/dailydotdev/daily-api/pull/2092/files#r1703623882

File: src/common/users.ts (hunk 4 above)
Comment (omBratteng) [reply src/common/users.ts]: Changed the function to this ☝️

Comment (capJavert) [reply src/common/users.ts]: Looks better yes, thanks for the effort. I would personally maybe split or expressions in separate if's if we add more down the line, but good for this stage.

**Review by nensidosari — APPROVED** [review]
Was trying to find names for the return conditions but too hard 😅

File: src/common/users.ts (hunk 4 above)
Comment (omBratteng) [reply src/common/users.ts]: I wanted to extract them into separate constants, but I couldn't come up with any good names for them :triggered:
